File: token-procedure-archetype/src/main/resources/archetype-resources/src/main/java/__pluginName__/__pluginName__IntrospectionTokenProcedure.java
package ${package}.${pluginName};

import se.curity.identityserver.sdk.procedure.token.IntrospectionTokenProcedure;
import se.curity.identityserver.sdk.procedure.token.context.IntrospectionTokenProcedurePluginContext;
import se.curity.identityserver.sdk.web.ResponseModel;

import java.util.HashMap;

public final class ${pluginName}IntrospectionTokenProcedure implements IntrospectionTokenProcedure
{
    private final ${pluginName}TokenProcedureConfig _configuration;

    public ${pluginName}IntrospectionTokenProcedure(${pluginName}TokenProcedureConfig configuration)
    {
        _configuration = configuration;
    }

    @Override
    public ResponseModel run(IntrospectionTokenProcedurePluginContext context)
    {
        var responseData = new HashMap<String, Object>();
        var presentedToken = context.getPresentedToken();

        responseData.put("active", presentedToken.isActive());

        if (presentedToken.isActive())
        {
            responseData.putAll(presentedToken.getTokenData().asMap());
            responseData.put("token_type", presentedToken.getType());
            responseData.put("client_id", presentedToken.getTokenDelegation().getClientId());
            responseData.put("expired_scope", presentedToken.getExpiredScopes());
        }

        return ResponseModel.mapResponseModel(responseData);
    }
}

File: kotlin-token-procedure-archetype/src/main/resources/archetype-resources/src/main/kotlin/__pluginName__/__pluginName__IntrospectionTokenProcedure.kt
package ${package}.${pluginName}

import se.curity.identityserver.sdk.procedure.token.IntrospectionTokenProcedure
import se.curity.identityserver.sdk.procedure.token.context.IntrospectionTokenProcedurePluginContext
import se.curity.identityserver.sdk.web.ResponseModel

class ${pluginName}IntrospectionTokenProcedure(private val _configuration: ${pluginName}TokenProcedureConfig): IntrospectionTokenProcedure
{

    override fun run(context: IntrospectionTokenProcedurePluginContext): ResponseModel
    {
        val presentedToken = context.presentedToken
        val responseData = mutableMapOf<String, Any>(
            "active" to presentedToken.isActive
        )

        if (presentedToken.isActive)
        {
            responseData.putAll(presentedToken.tokenData.asMap())
            responseData["token_type"] = presentedToken.type
            responseData["client_id"] = presentedToken.tokenDelegation.clientId
            responseData["expired_scope"] = presentedToken.expiredScopes
        }

        return ResponseModel.mapResponseModel(responseData)
    }
}
